velox/functions/prestosql/aggregates/MaxSizeForStatsAggregate.cpp

```diff
@@ -0,0 +1,269 @@
+      override {
+    BaseAggregate::doExtractValues(groups, numGroups, result, [&](char* group) {
+      return *BaseAggregate::Aggregate::template value<int64_t>(group);
+    });
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    for (auto i : indices) {
+      *BaseAggregate ::value<int64_t>(groups[i]) = 0;
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    // partial and final aggregations are the same
+    extractValues(groups, numGroups, result);
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::template updateGroups<true, int64_t>(
+        groups,
+        rows,
+        args[0],
+        [](int64_t& result, int64_t value) {
+          if (result < value) {
+            result = value;
+          }
+        },
+        mayPushdown);
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::updateOneGroup(
+        group,
+        rows,
+        args[0],
+        [](int64_t& result, int64_t value) {
+          result = result > value ? result : value;
+        },
+        [](int64_t& result, int64_t value, int /* unused */) {
+          result = value;
+        },
+        mayPushdown,
+        (int64_t)0);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdateSingleGroup(group, rows, args[0]);
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdate(groups, rows, args[0]);
+  }
+
+ protected:
+  void upsertOneAccumulator(
+      char* const group,
+      const BaseVector& inputVector,
+      std::vector<vector_size_t>& rowSizes,
+      vector_size_t idx) {
+    if (inputVector.isNullAt(idx)) {
+      return;
+    }
+
+    // Clear null
+    clearNull(group);
+    // Set max(current, this)
+    int64_t& cur = *value<int64_t>(group);
+    cur = std::max(cur, (int64_t)rowSizes[idx]);
+  }
+
+  void
+  doUpdate(char** groups, const SelectivityVector& rows, const VectorPtr& arg) {
+    DecodedVector decoded(*arg, rows, true);
+    auto indices = decoded.indices();
+    auto baseVector = decoded.base();
+
+    // TODO: Add test to verify that constant mapping encoding works w.r.t idx.
+    if (decoded.isConstantMapping() && decoded.isNullAt(0)) {
+      // nothing to do; all values are nulls
+      return;
+    }
+
+    auto numRows = arg->size();
```

Comment:
The code to calculate serialized sizes is repeated twice. Let's extract it into a helper function and re-use.

Also, it would be more efficient to use member variables for elementSizes, elementSizePtrs and elementIndices. This would allow to allocate memory for these once and re-use it (resizing if necessary) instead of allocating again for each batch.

Reply:
This comment is still not addressed.